Fix pthread key cleanup with musl libc

lrstewart opened the pull request:
### Description of changes: 

Customers building with musl libc instead of glibc saw segfaults from https://github.com/aws/s2n-tls/commit/ba2ffa8c0a132ee064801a5d3b0811936d0548dd if they didn't actually call s2n_init / use s2n.

The libmusl implementation of pthread keys is slightly more complicated than the standard one. When creating a key, libmusl also does some initialization of a field called [tsd](https://git.musl-libc.org/cgit/musl/tree/src/internal/pthread_impl.h#n49): [link](https://git.musl-libc.org/cgit/musl/tree/src/thread/pthread_key_create.c#n37). When deleting a key, libmusl assumes that initialization has been done and dereferences tsd without a null check: [link](https://git.musl-libc.org/cgit/musl/tree/src/thread/pthread_key_create.c#n65).

To avoid the segfault, we shouldn't try to delete the pthread key if we never created it. I do this by checking that s2n was initialized.

### Testing:
I added a new test that builds with musl.
I did it in a fairly hacky way just using musl-gcc, but it does work and did reproduce the customer issue. I think it's probably sufficient for now, but if reviewers disagree we could go for more formal cross compilation or build an Alpine docker image or something.

The new batchspec would just be called from GeneralBatch.
Here's a "musl" build failure with the new buildspec: [link](https://us-west-2.console.aws.amazon.com/codesuite/codebuild/024603541914/projects/s2nGeneralBatch/batch/s2nGeneralBatch%3A9facd969-4fca-49f0-88f2-6a03d9594a85?region=us-west-2) (it's the last build)
And here's a "musl" build success with my fix applied: [link](https://us-west-2.console.aws.amazon.com/codesuite/codebuild/024603541914/projects/s2nGeneralBatch/batch/s2nGeneralBatch%3Af1f1ac5d-cd12-4df7-8ed7-151921729e09/builds?region=us-west-2) (it's the last build)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

maddeleine commented on utils/s2n_init.c:
```diff
@@ -133,3 +133,8 @@ static void s2n_cleanup_atexit(void)
     (void) s2n_cleanup_atexit_impl();
 }
+
+bool s2n_is_init(void)
```
```suggestion
bool s2n_is_initialized(void)
```
Why not?

lrstewart replied:
I was trying to match `s2n_init`, but I can do the full word.

maddeleine commented on tests/unit/s2n_noop_test.c:
```diff
@@ -0,0 +1,24 @@
+#include "s2n_test.h"
+
+int main()
+{
```
Maybe explain in a comment why we have this test? Otherwise it looks really weird.

maddeleine commented:
Why doesn't your test build show the new s2n_noop_test failing?

lrstewart commented:
> Why doesn't your test build show the new s2n_noop_test failing?

Because while cleaning up I apparently broke the test :( But it probably shouldn't actually be its own standalone test. I've now moved it to s2n_random_test. Here's a failing build: https://us-west-2.console.aws.amazon.com/codesuite/codebuild/024603541914/projects/s2nGeneralBatch/build/s2nGeneralBatch%3A232dfe1a-1637-4b4e-a60e-4a0f55f3d258?region=us-west-2 You can see the changes I made to make it fail again in [this commit](https://github.com/aws/s2n-tls/commit/0a4b909afa833f4fdac5927e916000b1d7d1bf73). I also removed all the other test cases in s2n_random_test to prove that the new one triggers the error. Locally, I also proved it passes if the only test case is the basic "Random API" one instead.

goatgoose commented on tests/unit/s2n_random_test.c:
```diff
@@ -765,6 +765,17 @@ static int s2n_random_test_case_failure_cb(struct random_test_case *test_case)
     return EXIT_SUCCESS;
 }
 
+static int s2n_random_noop_destructor_test_cb(struct random_test_case *test_case)
```
Should this file be `s2n_fork_test` or something so it's less weird if we want to add other kinds of forking tests here, like this new one?

lrstewart replied:
No, this test is still specifically testing s2n_random.c. The problematic destructor was in s2n_random.c. That's why I put this new test case in s2n_random_test.c :)
